DATE_TIME should work as the primary time column for Pinot tables

npawar opened the pull request:
https://github.com/apache/incubator-pinot/issues/2756

This PR ensures that Pinot can use a DateTimeFieldSpec as a primary time column for the table.

After this change, we no longer need to use TimeFieldSpec. TimeFieldSpec, TimeGranularitySpec have been marked as Deprecated.
Some tests have been modified to test both cases, some have been changed to use datetime, and some retain time.

Manually tested creating segments, querying and retention for
1. Schema with multiple dateTimeFieldSpecs
2. Schema with just timeFieldSpec
3. Schema with no time
4. Schema with both time and dateTimeFieldSpec
5. Above all for hybrid tables as well (except the case with no time)

**Some changes**
1. The `segment.time.column.name` property in the segment metadata is now used to keep **primary time column**. This will match with the time column name in tableConfig. **This can be either timeFieldSpec or dateTimeFieldSpec**. 
If the primary time column is dateTimeFieldSpecs, the segment metadata will look like this
E.g: schema has dimensions `d1, d2,` metric `m1`, dateTimeFieldSpecs `dt1,dt2`.
Primary time = `dt1`
```
segment.dimension.column.names = d1,d2
segment.metric.column.names = m1
segment.datetime.column.names = dt1,dt2
segment.time.column.name = dt1 // Note the repetition
```
If the primary time column is timeFieldSpecs, the segment metadata will look like this
E.g: schema has dimensions `d1, d2,` metric `m1`, dateTimeFieldSpecs `dt1,dt2`, timeFieldSpec `t`
Primary time = `t`
```
segment.dimension.column.names = d1,d2
segment.metric.column.names = m1
segment.datetime.column.names = dt1,dt2
segment.time.column.name = t
```

haibow commented:
This diff looks like a milestone for TimeFieldSpec deprecation. @npawar shall we include this in 0.4.0 release? We could probably hold off on the release a bit. cc @kishoreg

haibow commented on pinot-broker/src/test/java/org/apache/pinot/broker/routing/timeboundary/TimeBoundaryManagerTest.java:
```diff
@@ -75,85 +75,112 @@ public void testTimeBoundaryManager() {
+  private void setSchemaDateTimeFieldSpec(String rawTableName, TimeUnit timeUnit) {
+    ZKMetadataProvider.setSchema(_propertyStore,
+        new Schema.SchemaBuilder().setSchemaName(rawTableName)
+            .addDateTime(TIME_COLUMN, FieldSpec.DataType.LONG, "1:"+timeUnit+":EPOCH", "1:"+timeUnit)
```
nit: please reformat this file

haibow commented on pinot-broker/src/main/java/org/apache/pinot/broker/routing/timeboundary/TimeBoundaryManager.java:
```diff
@@ -34,6 +34,8 @@
 import org.apache.pinot.spi.config.table.TableConfig;
 import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.data.DateTimeFieldSpec;
+import org.apache.pinot.spi.data.DateTimeFormatSpec;
```
Please format this file. Has unused imports.

haibow commented on pinot-core/src/main/java/org/apache/pinot/core/data/function/FunctionEvaluatorFactory.java:
```diff
@@ -59,7 +59,8 @@ public FunctionEvaluator getExpressionEvaluator(FieldSpec fieldSpec) {
     } else if (fieldSpec.getFieldType().equals(FieldSpec.FieldType.TIME)) {
 
-      // for backward compatible handling of TIME field conversion
+      // Time conversions should be done using DateTimeFieldSpec and treansformFunctions
```
nit: s/treansformFunctions/transformFunctions

npawar commented:
> This diff looks like a milestone for TimeFieldSpec deprecation. @npawar shall we include this in 0.4.0 release? We could probably hold off on the release a bit. cc @kishoreg

Yes, this is a milestone for TimeFieldSpec deprecation. Would be great if we can include in the release

mcvsubbu commented:
you have mentioned that you tested schema with no time on hybrid tables. How does that work for query routing?

npawar commented:
> you have mentioned that you tested schema with no time on hybrid tables. How does that work for query routing?

updated comment

mcvsubbu commented on pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:
```diff
@@ -507,10 +507,10 @@ public SchemaBuilder addMetric(String metricName, DataType dataType, Object defa
-     * TODO: This is going to be deprecated in favor of addDateTime().
-     *  Many tests use this to construct Schema with TimeFieldSpec.
-     *  This will continue to exist for a while, as it helps to test backward compatibility of schemas containing TimeFieldSpec
+     * Deprecated in favor of addDateTime().
```
```suggestion
     * @deprecated in favor of {@link #addDateTime()}.
```

mcvsubbu commented on pinot-spi/src/main/java/org/apache/pinot/spi/data/TimeFieldSpec.java:
```diff
@@ -28,6 +28,12 @@
 @JsonIgnoreProperties(ignoreUnknown = true)
+@Deprecated
+/**
+ * TimeFieldSpec is deprecated. Use {@link DateTimeFieldSpec} instead.
```
```suggestion
 * @deprecated TimeFieldSpec is deprecated. Use {@link DateTimeFieldSpec} instead.
```

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -362,12 +362,14 @@ private void writeMetadata()
     properties.setProperty(DATETIME_COLUMNS, config.getDateTimeColumnNames());
-    properties.setProperty(TIME_COLUMN_NAME, config.getTimeColumnName());
+    String timeColumnName = config.getTimeColumnName();
+    if (timeColumnName != null && !config.getSchema().getDateTimeNames().contains(timeColumnName)) {
```
Why do we need the second part of the check? What if the schema has no DateTimeField but only TimeField?

npawar replied:
Summarizing our offline discussion:
The second part was to avoid setting the time column name in `segment.datetime.column.names` and also in `segment.time.column.name`, which will happen if the primary time column is **dateTimeFieldSpec**. This happens because `segment.time.column.name` is **serving as primary timeColumnName and also as timeFieldSpec.** This worked well so far, because primary time column and timeFieldSpec were one and the same. 
But based on our chat, removing that check. As decided, the `segment.time.column.name` property will serve as the primary time column name. It will match with timecolumnName in tableConfig. It can be of type DATE_TIME or TIME.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -362,12 +362,14 @@ private void writeMetadata()
     // Write time related metadata (start time, end time, time unit)
-    String timeColumn = config.getTimeColumnName();
-    ColumnIndexCreationInfo timeColumnIndexCreationInfo = indexCreationInfoMap.get(timeColumn);
+    ColumnIndexCreationInfo timeColumnIndexCreationInfo = indexCreationInfoMap.get(timeColumnName);
```
Check whether `timeColumnName` is `null` first

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/segment/name/NormalizedDateSegmentNameGenerator.java:
```diff
@@ -60,16 +61,16 @@ public NormalizedDateSegmentNameGenerator(String tableName, @Nullable String seg
+      // Parse input time format: 'EPOCH' or 'SIMPLE_DATE_FORMAT' using pattern
+      Preconditions.checkNotNull(dateTimeFormatSpec);
+      TimeFormat timeFormat = dateTimeFormatSpec.getTimeFormat();
+      if (timeFormat.equals(TimeFormat.EPOCH)) {
```
```suggestion
      if (timeFormat == TimeFormat.EPOCH) {
```

mcvsubbu commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/V1Constants.java:
```diff
@@ -51,6 +51,10 @@
       public static final String METRICS = "segment.metric.column.names";
+      /**
+       * The primary time column for the table. This will match the timeColumnName defined in the tableConfig.
+       * In the Pinot schema, this column can be defined as either a TimeFieldSpec or DateTimeFieldSpec
```
mention in the comment that TimeFieldSpec is deprecated

mcvsubbu commented:
If you can please do these steps to test it
* Create a hybrid table with pinot-0.3.0, let it consume some segments, and have some offline segments as well
* Upgrade to the current version and then do the following tests:
1. Schema upgrades fine with a new field added (we are still able to get schema, etc.)
2. queries work fine (show data from old and new segments), even after new offline and realtime segments are pushed

Ideally, do the tests after upgrade controller, and then again after upgrading broker, and then after upgrading server

npawar commented:
> If you can please do these steps to test it
> 
> * Create a hybrid table with pinot-0.3.0, let it consume some segments, and have some offline segments as well
> * Upgrade to the current version and then do the following tests:
> 
> 1. Schema upgrades fine with a new field added (we are still able to get schema, etc.)
> 2. queries work fine (show data from old and new segments), even after new offline and realtime segments are pushed
> 
> Ideally, do the tests after upgrade controller, and then again after upgrading broker, and then after upgrading server

Just completed the following test. It all looks green.

Initial setup:
1. Started ZK and Kafka. Started Controller, Broker, Server using 0.3.0
2. Created kafka topic and hybrid table
3. Pushed 2 offline segments. Pushed data to kafka, allowed a segment to complete.
4. Checked getSchema, checked query, checked data correctness.
5. Added a field to schema. Reloaded segments. Restarted server.
6.  Checked getSchema, checked query, checked data correctness.

Controller upgrade:
1. Upgraded controller to new code
2. Checked getSchema, checked query, checked data correctness.
3. Added a field to schema. Reloaded segments. Restarted server.
4. Checked getSchema, checked query, checked data correctness.
5. Pushed one offline segment. Pushed some more realtime data letting more segments complete.
6. Checked getSchema, checked query, checked data correctness.

Broker upgrade:
1. Upgraded broker to new code
2. Checked getSchema, checked query, checked data correctness.
3. Added a field to schema. Reloaded segments. Restarted server.
4. Checked getSchema, checked query, checked data correctness.
5. Pushed one offline segment. Pushed some more realtime data letting more segments complete.
6. Checked getSchema, checked query, checked data correctness.

Server upgrade:
1. Upgraded server to new code
2. Checked getSchema, checked query, checked data correctness.
3. Added a field to schema. Reloaded segments. Restarted server.
4. Checked getSchema, checked query, checked data correctness.
5. Pushed one offline segment. Pushed some more realtime data letting more segments complete.
6. Checked getSchema, checked query, checked data correctness.

mcvsubbu commented:
Really appreciate your diligence @npawar thanks
